cpp/references/data_structures/list_digraph.cpp

```diff
@@ -3,636 +3,570 @@
-	Vector<VertexID> ret;
+Array ListDigraph::add_vertices(const Array &p_data, const bool &p_return) {
+	Array ret;
 	for (int i = 0; i < p_data.size(); i++) {
-		LemonNode vertex = _graph.addNode();
-		_vertex_map.set(vertex, p_data[i]);
+		LemonDigraph::Node node = _graph.addNode();
+		_vertex_map[node].data = p_data[i];
 
-		ret.push_back(LemonDigraph::id(vertex));
+		if (p_return)
+			ret.append(get_instance(node));
 	}
 	return ret;
 }
+Array ListDigraph::_add_vertices(Variant p_data, Variant p_return) {
+	ERR_FAIL_COND_V(p_data.get_type() != Variant::ARRAY, Array());
+	ERR_FAIL_COND_V(p_return.get_type() != Variant::BOOL, Array());
+	return add_vertices(p_data, p_return);
+}
 
-void ListDigraph::erase_vertex(const VertexID p_vertex) {
-	LemonNode node = LemonDigraph::nodeFromId(p_vertex);
-	ERR_FAIL_COND(!_graph.valid(node));
-	_graph.erase(node);
+void ListDigraph::_erase_vertex(Variant p_vertex) {
+	ERR_FAIL_COND(p_vertex.get_type() != Variant::OBJECT);
+	erase_vertex(p_vertex);
 }
 
-void ListDigraph::erase_vertices(Vector<VertexID> p_vertices) {
+void ListDigraph::erase_vertices(const Array &p_vertices) {
 	for (int i = 0; i < p_vertices.size(); i++) {
-		LemonNode node = LemonDigraph::nodeFromId(p_vertices[i]);
-		ERR_FAIL_COND(!_graph.valid(node));
-		_graph.erase(node);
+		Ref<Vertex> vertex = p_vertices[i];
+		ERR_CONTINUE(!vertex->is_valid());
+		_graph.erase(vertex->get_handle());
 	}
 }
+void ListDigraph::_erase_vertices(Variant p_vertices) {
+	ERR_FAIL_COND(p_vertices.get_type() != Variant::ARRAY);
+	erase_vertices(p_vertices);
+}
 
-void ListDigraph::reserve_vertices(int p_amount) {
-	_graph.reserveNode(p_amount);
+void ListDigraph::_reserve_vertices(Variant p_amount) {
+	ERR_FAIL_COND(p_amount.get_type() != Variant::INT);
+	reserve_vertices(p_amount);
 }
 
-Vector<VertexID> ListDigraph::get_vertices() {
-	Vector<VertexID> ret;
-	for (LemonNodeIt it(_graph); it != lemon::INVALID; ++it) {
-		ret.push_back(LemonDigraph::id(it));
+Array ListDigraph::_get_vertices() {
+	Array ret;
+	for (LemonDigraph::NodeIt it(_graph); it != lemon::INVALID; ++it) {
+		ret.append(get_instance(it));
 	}
 	return ret;
 }
 
-ArcID ListDigraph::add_arc(VertexID p_source, VertexID p_target) {
-	LemonNode source = LemonDigraph::nodeFromId(p_source);
-	ERR_FAIL_COND_V(!_graph.valid(source), -1);
-
-	LemonNode target = LemonDigraph::nodeFromId(p_target);
-	ERR_FAIL_COND_V(!_graph.valid(target), -1);
-
-	LemonArc arc = _graph.addArc(source, target);
-	return LemonDigraph::id(arc);
+Ref<Arc> ListDigraph::_add_arc(Variant p_source, Variant p_target) {
+	ERR_FAIL_COND_V(p_source.get_type() != Variant::OBJECT, nullptr);
+	ERR_FAIL_COND_V(p_target.get_type() != Variant::OBJECT, nullptr);
+	return add_arc(p_source, p_target);
 }
 
-Vector<ArcID> ListDigraph::add_arcs(Vector<HashMap<String, Variant>> p_data) {
-	Vector<ArcID> ret;
+Array ListDigraph::add_arcs(const Array &p_data, const bool &p_return) {
+	Array ret;
```

Comment:
By large, these batch methods are very, very similar to each other (between Array versus Vector<T> and Dictionary versus HashMap<K, V>). I think we can consolidate things even further if we developed took the differences in their API and then created template methods that wrapped the inconsistent parts to create a unified API.

So, rather than having `!dict.has("source")` for Dictionaries and `dict.count("source") <= 0` for HashMap, we can just have...

    template <typename T>
    bool has_key(const T &map, const String &p_key) {
        return map.has(p_key);
    }
    template <>
    bool has_key<HashMap>(const HashMap &map, const String &p_key) {
        return m.count(p_key) >= 0;
    }

Then, when you implement `add_arcs`, you can just have one copy that uses `has_key(p_data[i])` instead. Another approach that might(?) work would be to use a pointer to a HashMap and just one method that takes a Variant and checks for the type. Either way, I think something of this nature could help merge together a lot of the split up batch operations we have currently, since most of the logic is identical. We would just need to create wrapper methods for the portions that are different (getting the data structure size for the loop, checking if a record exists, getting a record, setting a record, etc.).

Reply:
My idea was to avoid calling methods inside the batch methods.... because as you said batch methods should be as performant as possible... calls to other methods will lower that performance a bit (that was the only reason I implemented it this way) :)

Reply:
> My idea was to avoid calling methods inside the batch methods...calls to other methods will lower that performance a bit

1. Calls to other functions won't be that big of a deal to the performance, so long as the added functions are C++ functions and not scripting API methods.
2. We are already calling scripting API methods when we need to be calling scripting API methods. Creating wrappers around those methods to unify the scripting vs. C++ variations won't increase the number of scripting API calls made, but will instead just add one C++ call. And as mentioned in point 1, those won't have that big of an impact.
3. The main performance improvement we get from batch methods is from not having to make N scripting API calls for the graph methods, like insert, merge, set, get, has, etc. all wrapped in a for loop on the scripting API side (which further exacerbates things). When we hand off a batch job to one method, and run the loop *inside* that method, then even if we have to run scripting API calls, it will be happening entirely from the C++ side and not between C++ and GDScript or some other language.